lib/auth/init_test.go

```diff
@@ -158,85 +158,124 @@ func TestBadIdentity(t *testing.T) {
 // TestAuthPreference ensures that the act of creating an AuthServer sets
 // the AuthPreference (type and second factor) on the backend.
 func TestAuthPreference(t *testing.T) {
-	tempDir := t.TempDir()
-
-	bk, err := lite.New(context.TODO(), backend.Params{"path": tempDir})
-	require.NoError(t, err)
-
 	ap, err := services.NewAuthPreference(services.AuthPreferenceSpecV2{
-		Type:         "local",
-		SecondFactor: "u2f",
+		Type:         constants.Local,
+		SecondFactor: constants.SecondFactorU2F,
 		U2F: &services.U2F{
 			AppID:  "foo",
 			Facets: []string{"bar", "baz"},
 		},
 	})
 	require.NoError(t, err)
 
-	clusterName, err := services.NewClusterName(services.ClusterNameSpecV2{
-		ClusterName: "me.localhost",
-	})
-	require.NoError(t, err)
-
-	staticTokens, err := services.NewStaticTokens(services.StaticTokensSpecV2{
-		StaticTokens: []services.ProvisionTokenV1{},
-	})
-	require.NoError(t, err)
-
-	ac := InitConfig{
-		DataDir:        tempDir,
-		HostUUID:       "00000000-0000-0000-0000-000000000000",
-		NodeName:       "foo",
-		Backend:        bk,
-		Authority:      testauthority.New(),
-		ClusterConfig:  services.DefaultClusterConfig(),
-		ClusterName:    clusterName,
-		StaticTokens:   staticTokens,
-		AuthPreference: ap,
-	}
-	as, err := Init(ac)
+	conf := setupConfig(t)
+	conf.AuthPreference = ap
+	as, err := Init(conf)
 	require.NoError(t, err)
 	defer as.Close()
 
 	cap, err := as.GetAuthPreference()
 	require.NoError(t, err)
-	require.Equal(t, cap.GetType(), "local")
+	require.Equal(t, cap.GetType(), constants.Local)
 	require.Equal(t, cap.GetSecondFactor(), constants.SecondFactorU2F)
-
 	u, err := cap.GetU2F()
 	require.NoError(t, err)
 	require.Equal(t, u.AppID, "foo")
 	require.Equal(t, u.Facets, []string{"bar", "baz"})
 }
 
-func TestClusterID(t *testing.T) {
-	tempDir := t.TempDir()
+// TestAuthPreferenceResetConfigFileToDefaults ensures that a config-file
+// AuthPreference can be reset back to defaults.
+func TestAuthPreferenceResetConfigFileToDefaults(t *testing.T) {
+	// Simulate a server with auth preference from config file.
+	var err error
+	conf := setupConfig(t)
+	conf.AuthPreference, err = services.NewAuthPreferenceFromConfigFile(services.AuthPreferenceSpecV2{
+		SecondFactor: constants.SecondFactorOff,
+	})
+	require.NoError(t, err)
+	authServer, err := Init(conf)
+	require.NoError(t, err)
+	defer authServer.Close()
 
-	bk, err := lite.New(context.TODO(), backend.Params{"path": tempDir})
+	// Verify the stored auth preference is labelled as originating from config
+	// file.
+	storedAuthPref, err := authServer.GetAuthPreference()
 	require.NoError(t, err)
+	require.True(t, storedAuthPref.IsFromConfigFile())
+	defaultSecondFactor := services.DefaultAuthPreference().GetSecondFactor()
+	require.NotEqual(t, defaultSecondFactor, storedAuthPref.GetSecondFactor())
```

Comment:
nit: do `cmp.Diff` on the whole `AuthPreference`, here and below

Reply:
That doesn't really work because of the metadata changes that are involved.

Reply:
Are there many metadata changes?
You can use https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#IgnoreFields to skip stuff like timestamps